src/triage/component/catwalk/predictors.py

```diff
@@ -87,6 +88,34 @@ def _existing_predictions(self, Prediction_obj, session, model_id, matrix_store)
             .filter(Prediction_obj.as_of_date.in_(self._as_of_dates(matrix_store)))
         )
 
+    @db_retry
+    def needs_predictions(self, matrix_store, model_id):
+        """Returns whether or not the given matrix and model are lacking any predictions
+
+        Args:
+            matrix_store (triage.component.catwalk.storage.MatrixStore) A matrix with metadata
+            model_id (int) A database ID of a model
+
+        The way we check is by grabbing all the distinct as-of-dates in the predictions table
+        for this model and matrix. If there are more as-of-dates defined in the matrix's metadata
+        than are in the table, we need predictions
+        """
+        if not self.save_predictions:
+            return False
+        session = self.sessionmaker()
+        prediction_obj = matrix_store.matrix_type.prediction_obj
+        as_of_dates_in_db = set(
+            obj.as_of_date.date()
+            for obj in session.query(prediction_obj).filter_by(
+                model_id=model_id,
+                matrix_uuid=matrix_store.uuid
+            ).distinct(prediction_obj.as_of_date).all()
+        )
```

Comment:
I think this could be a tiny bit simpler / faster with the `value()` method, _e.g._:

```suggestion
        as_of_dates_in_db = {
            as_of_date.date()
            for as_of_date in session.query(prediction_obj).filter_by(
                …
            ).distinct(prediction_obj.as_of_date).value('as_of_date')
        }
```

(…Or using whatever the column expression/label would be in that context.)

(And note that `Query.all()` is just syntactic sugar, the same as `list(query)` – that is, it's for when you need a list of results, as opposed to some other case, as here.)

:man_shrugging:

Reply:
So one of my test cases fails when I make this change with 'NoneType is not iterable'. I think this means that .value('as_of_date') doesn't work how we need it to when there are no rows

Reply:
I _think_ I just typo'd the method name. Iteration of a query should never simply return `None` -- it should always iterate.

I believe the correct method is `values(*columns)` [[code]](https://github.com/sqlalchemy/sqlalchemy/blob/c0f9708fde39175c8695fdd87873464f516fdf98/lib/sqlalchemy/orm/query.py#L1361). So it would just be:

```
        as_of_dates_in_db = {
            as_of_date.date()
            for (as_of_date,) in session.query(prediction_obj).filter_by(
                model_id=model_id,
                matrix_uuid=matrix_store.uuid
            ).distinct(prediction_obj.as_of_date).values('as_of_date')
        }
```

(`value(column)` looks to be a helper wrapper of `values(*columns)`, which returns the first column of the first row of a query result -- so I guess useful when you're doing aggregations, for example. That's why it falls back to `None`.)